x/authority/keeper/keeper.go

```diff
@@ -42,31 +45,61 @@ func NewKeeper(cdc codec.BinaryMarshaler, storeKey sdk.StoreKey, issuerKeeper is
 	}
 }
 
-func (k Keeper) SetAuthority(ctx sdk.Context, authority sdk.AccAddress) {
-	store := ctx.KVStore(k.storeKey)
+// BootstrapAuthority solely exists for the genesis establishment of the chain
+// authority. Once the authority is set, invoking this function again will panic.
+func (k Keeper) BootstrapAuthority(ctx sdk.Context, newAuthority sdk.AccAddress) {
+	authorityAcc, _, _ := k.getAuthority(ctx)
 
-	if store.Get([]byte(keyAuthorityAccAddress)) != nil {
-		panic("Authority was already specified")
+	// set authority only if it is not set up.
+	if !authorityAcc.Empty() {
+		panic(errors.New("authority is set and sealed"))
 	}
 
-	bz := k.cdc.MustMarshalBinaryBare(&types.Authority{Address: authority.String()})
-	store.Set([]byte(keyAuthorityAccAddress), bz)
+	k.saveAuthorities(ctx, newAuthority, "")
 }
 
-func (k Keeper) GetAuthority(ctx sdk.Context) sdk.AccAddress {
+func (k Keeper) saveAuthorities(
+	ctx sdk.Context, newAuthority sdk.AccAddress, formerAuthorityAddr string,
+) {
 	store := ctx.KVStore(k.storeKey)
-	bz := store.Get([]byte(keyAuthorityAccAddress))
-	var authority types.Authority
-	k.cdc.MustUnmarshalBinaryBare(bz, &authority)
-	acc, err := sdk.AccAddressFromBech32(authority.Address)
+
+	bz := k.cdc.MustMarshalBinaryBare(
+		&types.Authority{
+			Address:       newAuthority.String(),
+			FormerAddress: formerAuthorityAddr,
+			LastModified:  ctx.BlockTime(),
+		},
+	)
+	store.Set([]byte(keyAuthorityAccAddress), bz)
+}
+
+func (k Keeper) getAuthority(ctx sdk.Context) (authority sdk.AccAddress, formerAuthority sdk.AccAddress, err error) {
+	authoritySet := k.GetAuthoritySet(ctx)
+	authority, err = sdk.AccAddressFromBech32(authoritySet.Address)
 	if err != nil {
-		panic(err.Error())
+		return nil, nil, err
+	}
+
+	if authoritySet.LastModified.Add(types.AuthorityTransitionDuration).After(ctx.BlockTime()) {
+		// within the transition period we keep the former address
+		formerAuthority, _ = sdk.AccAddressFromBech32(authoritySet.FormerAddress)
 	}
-	return acc
+
+	return authority, formerAuthority, nil
+}
+
+func (k Keeper) GetAuthoritySet(ctx sdk.Context) types.Authority {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get([]byte(keyAuthorityAccAddress))
+	var authoritySet types.Authority
+	k.cdc.MustUnmarshalBinaryBare(bz, &authoritySet)
+	return authoritySet
 }
```

Comment:
I would prefer this function to be merged into `getAuthority` and then possible have `getAuthority` be made public if necessary.

Reply:
- We need an exported function for the genesis export function and the formerAuth business rules of `GetAuthority()` are unnecessary for the Genesis func.
- After merging the state retrieval code with higher business rules, it looks a bit long after following your next great suggestion which I like a lot :)

Reply:
Can't the genesis export code use the `getAuthorities` function? It only needs the current authority account.

Reply:
Sure it could, but this is sufficient to what it needs. The getAuthorities function has the former Authority related timestamp business logic that is irrelevant to the Genesis function.